### PR TITLE
`spin new`: add component to existing application

### PR DESCRIPTION
**Pull request by itowlson**
This updates the templates system to allow for adding a component to an existing application as well as creating a new application.

Here's how it looks in use:

```
ivan@hecate:~/testing$ spin new http-empty compytest
Project description:
HTTP base: /

ivan@hecate:~/testing$ cd compytest/

ivan@hecate:~/testing/compytest$ spin new http-rust hello
Project description:
HTTP path: /hello

ivan@hecate:~/testing/compytest$ spin new http-rust goodbye
Project description:
HTTP path: /goodbye

ivan@hecate:~/testing/compytest$ code .
# ... build spiffy application ...

ivan@hecate:~/testing/compytest$ spin build
# ... Rust compiler spew ...

ivan@hecate:~/testing/compytest$ spin up
Serving http://127.0.0.1:3000
Available Routes:
  hello: http://127.0.0.1:3000/hello
  goodbye: http://127.0.0.1:3000/goodbye
```

The template has to support the feature - so far I've only done it for `http-rust` but if we're happy with the design then I'll do it for the others.

A few things that I'd particularly keen to hear thoughts on:

1. Should we try to make it so this Just Works without requiring the incantations in the templates?  This would simplify template development and maintenance but would require much more knowledge of Spin file formats baked into the templating code.
2. Is there a simpler way to approach the structure?  I want to re-assess where various bits of code have landed but at the moment I feel l have too many things doing similar jobs or with similar names.
3. Does the "add or create new" inference look right or should I provide an explicit override a la `spin new --add`?

Thanks!

ETA: oh yeah I need to write some tests too

### CODE REVIEW

File: crates/templates/src/lib.rs
Comment (itowlson) [inline crates/templates/src/lib.rs:3]: This is because I'm not sure of the interface and I want to wait for feedback before writing docs... please help me remember to put it back when things firm up!

**Comment by mikkelhegn** [comment]
> Should we try to make it so this Just Works without requiring the incantations in the templates? This would simplify template development and maintenance but would require much more knowledge of Spin file formats baked into the templating code.

From looking at the PR code, I think having the component-snippet as part of the template opens up for more template scenarios. E.g., a template for a file-server, which has specific configuration in the component. So I don't think this is a bad idea.

> Does the "add or create new" inference look right or should I provide an explicit override a la spin new --add?

I think adding something to an existing structure should be explicit, and different than new. The thing that always throws me off with `spin new` today is the directories being created. E.g., what does `new` do if I'm in a directory with a `spin.toml`, but I'm using `--output` to point to a subdirectory? That confusion might be introduced if it's now explicit - e.g., `spin add` or `spin component add` (There's now a noun-verb thing in there 😂)

**Comment by kate-goldenring** [comment]
How do we prevent or inform a user that they cannot create an app with components of different triggers? Would this be caught during `spin new` (or `spin add` if named as such)?

**Comment by fibonacci1729** [comment]
This is great @itowlson ! I ran through your example and it works as you mention. A few thoughts.

From the perspective of the user crafting a new application, the effects of `spin new ...` (as implemented here), are dictated by the presence of `spin.toml`, i.e. whether or not we are adding components from snippets or drafting a new application is inferred from the presence of a `spin.toml`. I think I agree with Mikkel's suggestion that teasing out this implicit-ness into an `add` command is the best UX.

It would be great to give "snippets" a name so you can say something like: 
```
spin new http-empty myapp && cd myapp
spin add redirect fileserver
```
OR (as a single command)
```
spin new http-empty myapp --add redirect fileserver
```

... Where `fileserver` and `redirect` are the explicit template authored names given to the component snippets that can be added.

I'm probably missing some nuance that you've already considered so apologies if my suggestions are dimwitted.

**Comment by kate-goldenring** [comment]
> 1. Should we try to make it so this Just Works without requiring the incantations in the templates? This would simplify template development and maintenance but would require much more knowledge of Spin file formats baked into the templating code.

I like the idea of keeping the details in the templates. Yes, more work for template developers, but ensures that Spin doesnt need to know language specific details

> 2. Is there a simpler way to approach the structure? I want to re-assess where various bits of code have landed but at the moment I feel l have too many things doing similar jobs or with similar names.

I am curious on how you see the snippets expanding to more than components. Also, have we ever considered having a spin config file? In the config, the user could specify "I want an app with 1 rust http-server named hello and 1 go http-server named goodbye". Then additions could happen like `helm upgrade` wherein an expanded config is applied. 
These can just be futuristic thoughts. I think this vastly simplifies the experience of adding a component to an existing app!

> 3. Does the "add or create new" inference look right or should I provide an explicit override a la spin new --add?

I agree with the previous recs to have an explicit `add` subcommand

**Comment by itowlson** [comment]
@kate-goldenring I woke up last night worrying about exactly the trigger mismatch scenario!  I can add some metadata to the template for that, hopefully in a nonbreaking way.  Thanks for catching this!

File: templates/http-rust/metadata/spin-template.toml
Comment (itowlson) [reply templates/http-rust/metadata/spin-template.toml:8]: I did wonder about that!  It might need a bit of extra magic to make the paths work but that should just be a Liquid filter.  I am a little concerned about that being a breaking change, though I think technically the new section is a breaking change anyway (our TOML reader is set to fail on unrecognised elements).  I will ponder this more - thanks!

**Comment by itowlson** [comment]
@fibonacci1729 I hadn't thought of `spin new foo --add bar` or `spin add bar quux` - thanks!  It would need a bit of rework to manage the multiple sets of prompts though (e.g. which `HTTP path` am I entering here), so might be something to defer to a subsequent enhancement?

**Comment by itowlson** [comment]
@kate-goldenring Could you amplify what you mean by _"In the config, the user could specify "I want an app with 1 rust http-server named hello and 1 go http-server named goodbye". Then additions could happen like helm upgrade wherein an expanded config is applied."_?  Are you describing expansion at runtime/deploy time?

**Comment by kate-goldenring** [comment]
> @kate-goldenring Could you amplify what you mean by _"In the config, the user could specify "I want an app with 1 rust http-server named hello and 1 go http-server named goodbye". Then additions could happen like helm upgrade wherein an expanded config is applied."_? Are you describing expansion at runtime/deploy time?

This would be at application creation time. Instead of `spin new` and following CLI steps to create app, it would be `spin new --config app-config.toml` where the app config could specify all the different apps that should be a part of the Spin app. then, to add an app, the additional app would be added to `app-config.toml` and `spin upgrade --config app-config.toml` could be run. Spin reconciles the changes and adds the app to the config. May be messy.

**Comment by itowlson** [comment]
@kate-goldenring re "how you see the snippets expanding to more than components"... I don't really, as yet, but I don't want to bake my lack of imagination into the code, and I think the cost of the YAGNI in this case is pretty small.

But... saying that... there might be an attraction to templates that could add, say, routing and form processing, if those were specific to Spin or could use the Spin config info.  Or longer term adding component dependencies, e.g. bolting a database or auth Wasm module onto the side of an application module a la sidecar.  But I'm hand-waving here.  I don't know how these would work, and it's quite possible the proposed structure wouldn't extend to those after all.

Hope that makes sense and I'd love to hear your thoughts!

**Comment by kate-goldenring** [comment]
> bolting a database or auth Wasm module onto the side of an application module a la sidecar

I wonder for more specific scenarios in the future if we will want to add the ability to add snippets via command line or even via spin plugins. but regardless sounds good to me to have a generic snippet concept and see how it fits feature requests in the future

**Comment by itowlson** [comment]
Thanks for all the feedback folks!

Regarding `new` vs `add`, I'll change this to `spin add <template> <name>`.  This also means we can have the `-f` syntax (same as `up`) to reference a non-default app manifest e.g. `spin add -f ui/spin.toml`.

I'm a little anxious this is going to start blowing up the `spin` subcommand list with template-related stuff... are folks okay with a top-level `spin add`?

**Comment by mikkelhegn** [comment]
I think for now this is ok, but I share your concern.

**Review by fibonacci1729 — APPROVED** [review]
LGTM

**Pull request merged**
